File: alpha/model_adapter.R
#' Model-Agnostic LLM Completion and Embeddings Adapter
#'
#' Unified wrapper for calling chat completion and embedding endpoints
#' across OpenRouter, Ollama, OpenAI, and Gemini.
#'
#' @importFrom httr2 request req_headers req_body_json req_perform resp_body_json req_retry
#' @importFrom jsonlite fromJSON
#' @export

#' Generate Chat Completion
#'
#' Call chat completion endpoint of the configured provider.
#'
#' @param prompt User prompt text.
#' @param system_prompt Optional system instructions.
#' @param json_mode Boolean, whether to request JSON response formatting.
#' @param config Configuration list containing api keys and model settings.
#' @return A list containing the generated text or parsed JSON.
generate_completion <- function(prompt, system_prompt = NULL, json_mode = FALSE, config = NULL) {
    if (is.null(config)) {
        stop("Configuration list must be provided.")
    }
    
    provider <- tolower(config$llm_provider)
    model <- config$llm_model
    
    # 1. Build messages structure
    messages <- list()
    if (!is.null(system_prompt) && system_prompt != "") {
        messages[[length(messages) + 1]] <- list(role = "system", content = system_prompt)
    }
    messages[[length(messages) + 1]] <- list(role = "user", content = prompt)
    
    content <- NULL
    reasoning <- NULL
    
    # 2. Execute call depending on provider
    if (provider == "openrouter") {
        if (config$openrouter_api_key == "") {
            stop("OPENROUTER_API_KEY is not configured.")
        }
        
        req <- httr2::request("https://openrouter.ai/api/v1/chat/completions") %>%
            httr2::req_headers(
                "Authorization" = paste("Bearer", config$openrouter_api_key),
                "HTTP-Referer" = "https://github.com/Arf9999/newsletter_phase2",
                "X-Title" = "Narrative Intelligence Pipeline"
            )
            
        body <- list(
            model = model,
            messages = messages
        )
        if (json_mode) {
            body$response_format <- list(type = "json_object")
        }
        
        req <- req %>% httr2::req_body_json(body) %>% httr2::req_retry(max_tries = 3)
        resp <- httr2::req_perform(req)
        res_json <- httr2::resp_body_json(resp)
        
        content <- res_json$choices[[1]]$message$content
        reasoning <- res_json$choices[[1]]$message$reasoning_content
        if (is.null(reasoning)) {
            reasoning <- res_json$choices[[1]]$message$reasoning
        }
        
    } else if (provider == "ollama") {
        url <- paste0(config$ollama_host, "/api/chat")
        body <- list(
            model = model,
            messages = messages,
            stream = FALSE
        )
        if (json_mode) {
            body$format <- "json"
        }
        
        req <- httr2::request(url) %>%
            httr2::req_body_json(body) %>%
            httr2::req_retry(max_tries = 2)
            
        resp <- httr2::req_perform(req)
        res_json <- httr2::resp_body_json(resp)
        
        content <- res_json$message$content
        
    } else if (provider == "openai") {
        if (config$openai_api_key == "") {
            stop("OPENAI_API_KEY is not configured.")
        }
        
        req <- httr2::request("https://api.openai.com/v1/chat/completions") %>%
            httr2::req_headers("Authorization" = paste("Bearer", config$openai_api_key))
            
        body <- list(
            model = model,
            messages = messages
        )
        if (json_mode) {
            body$response_format <- list(type = "json_object")
        }
        
        req <- req %>% httr2::req_body_json(body) %>% httr2::req_retry(max_tries = 3)
        resp <- httr2::req_perform(req)
        res_json <- httr2::resp_body_json(resp)
        
        content <- res_json$choices[[1]]$message$content
        reasoning <- res_json$choices[[1]]$message$reasoning_content
        
    } else if (provider == "gemini") {
        if (config$gemini_api_key == "") {
            stop("GEMINI_API_KEY is not configured.")
        }
        
        url <- paste0("https://generativelanguage.googleapis.com/v1beta/models/", model, ":generateContent?key=", config$gemini_api_key)
        
        # Build contents structure for Gemini
        contents <- list(
            parts = list(
                list(text = prompt)
            )
        )
        
        # System instructions if provided
        body <- list(contents = list(contents))
        if (!is.null(system_prompt) && system_prompt != "") {
            body$systemInstruction <- list(parts = list(list(text = system_prompt)))
        }
        if (json_mode) {
            body$generationConfig <- list(responseMimeType = "application/json")
        }
        
        req <- httr2::request(url) %>%
            httr2::req_body_json(body) %>%
            httr2::req_retry(max_tries = 3)
            
        resp <- httr2::req_perform(req)
        res_json <- httr2::resp_body_json(resp)
        
        content <- res_json$candidates[[1]]$content$parts[[1]]$text
        
        # Gemini thought extraction if reasoning model and parts contain thought field
        parts <- res_json$candidates[[1]]$content$parts
        if (length(parts) > 1) {
            for (i in seq_along(parts)) {
                if (!is.null(parts[[i]]$thought)) {
                    reasoning <- parts[[i]]$text
                    break
                }
            }
        }
        
    } else {
        stop("Unknown LLM provider specified: ", provider)
    }
    
    # 3. Log to Forensic Log
    log_forensic_response(
        provider = provider,
        model = model,
        prompt = prompt,
        system_prompt = system_prompt,
        response_text = content,
        reasoning_content = reasoning,
        config = config
    )
    
    return(content)
}

#' Generate Vector Embeddings
#'
#' Generate embeddings for a vector of texts using the configured provider.
#'
#' @param texts A character vector of texts to embed.
#' @param config Configuration list.
#' @param space Either "english" or "multilingual" to toggle models if needed.
#' @return A list of numeric vectors containing the embeddings.
generate_embeddings <- function(texts, config = NULL, space = "english") {
    if (is.null(config)) {
        stop("Configuration list must be provided.")
    }
    if (length(texts) == 0) {
        return(list())
    }
    
    provider <- tolower(config$embedding_provider)
    # Check if we should override model based on vector space target
    model <- config$embedding_model
    
    embeddings <- list()
    
    for (txt in texts) {
        if (txt == "" || is.na(txt)) {
            embeddings[[length(embeddings) + 1]] <- NA
            next
        }
        
        if (provider == "openrouter") {
            # OpenRouter typically doesn't host embeddings natively, so we fall back to OpenAI API format
            # or custom endpoint. Let's build OpenAI/OpenRouter fallback for embeddings.
            api_key <- ifelse(config$openrouter_api_key != "", config$openrouter_api_key, config$openai_api_key)
            if (api_key == "") {
                stop("Neither OPENROUTER_API_KEY nor OPENAI_API_KEY is configured for embeddings.")
            }
            
            # Use OpenAI embeddings endpoint (which many API hosts share)
            req <- httr2::request("https://api.openai.com/v1/embeddings") %>%
                httr2::req_headers("Authorization" = paste("Bearer", api_key)) %>%
                httr2::req_body_json(list(
                    model = model,
                    input = txt
                )) %>%
                httr2::req_retry(max_tries = 3)
                
            resp <- httr2::req_perform(req)
            res_json <- httr2::resp_body_json(resp)
            embeddings[[length(embeddings) + 1]] <- as.numeric(res_json$data[[1]]$embedding)
            
        } else if (provider == "openai") {
            if (config$openai_api_key == "") {
                stop("OPENAI_API_KEY is not configured.")
            }
            
            req <- httr2::request("https://api.openai.com/v1/embeddings") %>%
                httr2::req_headers("Authorization" = paste("Bearer", config$openai_api_key)) %>%
                httr2::req_body_json(list(
                    model = model,
                    input = txt
                )) %>%
                httr2::req_retry(max_tries = 3)
                
            resp <- httr2::req_perform(req)
            res_json <- httr2::resp_body_json(resp)
            embeddings[[length(embeddings) + 1]] <- as.numeric(res_json$data[[1]]$embedding)
            
        } else if (provider == "ollama") {
            url <- paste0(config$ollama_host, "/api/embed")
            req <- httr2::request(url) %>%
                httr2::req_body_json(list(
                    model = model,
                    input = txt
                )) %>%
                httr2::req_retry(max_tries = 2)
                
            resp <- httr2::req_perform(req)
            res_json <- httr2::resp_body_json(resp)
            embeddings[[length(embeddings) + 1]] <- as.numeric(res_json$embeddings[[1]])
            
        } else if (provider == "gemini") {
            if (config$gemini_api_key == "") {
                stop("GEMINI_API_KEY is not configured.")
            }
            
            # Use developer API embedContent
            url <- paste0("https://generativelanguage.googleapis.com/v1beta/models/", model, ":embedContent?key=", config$gemini_api_key)
            
            req <- httr2::request(url) %>%
                httr2::req_body_json(list(
                    content = list(
                        parts = list(
                            list(text = txt)
                        )
                    )
                )) %>%
                httr2::req_retry(max_tries = 3)
                
            resp <- httr2::req_perform(req)
            res_json <- httr2::resp_body_json(resp)
            embeddings[[length(embeddings) + 1]] <- as.numeric(res_json$embedding$values)
            
        } else {
            stop("Unknown embedding provider specified: ", provider)
        }
    }
    
    return(embeddings)
}

#' Log LLM response activity to a Forensic Log file
#'
#' Appends raw system prompts, user prompts, generated completions, and model thinking processes to a configured path.
#'
#' @param provider The backend provider name.
#' @param model The active model name.
#' @param prompt User prompt text.
#' @param system_prompt System prompt instructions.
#' @param response_text Final answer text.
#' @param reasoning_content Optional model reasoning/thinking process text.
#' @param config Config list.
log_forensic_response <- function(provider, model, prompt, system_prompt, response_text, reasoning_content = NULL, config = NULL) {
    log_path <- if (!is.null(config) && !is.null(config$forensic_log_path) && config$forensic_log_path != "") {
        config$forensic_log_path
    } else {
        "FORENSIC_LOG"
    }
    
    # Ensure parent directory exists if log is placed inside a folder path
    dir_name <- dirname(log_path)
    if (dir_name != "." && dir_name != "" && !dir.exists(dir_name)) {
        dir.create(dir_name, recursive = TRUE, showWarnings = FALSE)
    }
    
    timestamp <- format(Sys.time(), "%Y-%m-%d %H:%M:%S")
    
    system_prompt_str <- if (!is.null(system_prompt) && system_prompt != "") {
        paste0("SYSTEM PROMPT:\n", system_prompt, "\n")
    } else {
        ""
    }
    
    log_entry <- paste0(
        "================================================================================\n",
        "TIMESTAMP: ", timestamp, "\n",
        "PROVIDER:  ", provider, "\n",
        "MODEL:     ", model, "\n",
        system_prompt_str,
        "------------------------------------ PROMPT ------------------------------------\n",
        prompt, "\n",
        "---------------------------------- RESPONSE ------------------------------------\n"
    )
    
    if (!is.null(reasoning_content) && reasoning_content != "") {
        log_entry <- paste0(
            log_entry,
            "--- THINKING PROCESS ---\n",
            reasoning_content, "\n",
            "--- FINAL ANSWER ---\n"
        )
    }
    
    log_entry <- paste0(
        log_entry,
        response_text, "\n",
        "================================================================================\n\n"
    )
    
    # Append to the log file
    tryCatch({
        cat(log_entry, file = log_path, append = TRUE)
    }, error = function(e) {
        warning("Failed to write to FORENSIC_LOG: ", e$message)
    })
}

File: alpha/model_adapter.py
import os
import re
import json
import time
import httpx

def get_config():
    # Load credentials.json
    creds_path = "alpha/credentials.json"
    config = {
        "gmail_username": "",
        "gmail_app_password": "",
        "db_path": "alpha/newsletters.db",
        "llm_provider": "openrouter",
        "llm_model": "liquid/lfm-2-24b-a2b",
        "embedding_provider": "openrouter",
        "embedding_model": "nomic-embed-text",
        "openrouter_api_key": "",
        "ollama_host": "http://localhost:11434",
        "openai_api_key": "",
        "gemini_api_key": "",
        "forensic_log_path": "FORENSIC_LOG"
    }
    
    if os.path.exists(creds_path):
        try:
            with open(creds_path, "r") as f:
                creds = json.load(f)
                for k, v in creds.items():
                    # Map both original and lowercase keys
                    config[k.lower()] = v
                    config[k] = v
        except Exception as e:
            print(f"Warning: Failed to load credentials: {e}")
            
    # Load manifest.json
    manifest_path = "manifest.json"
    if os.path.exists(manifest_path):
        try:
            with open(manifest_path, "r") as f:
                manifest = json.load(f)
                models = manifest.get("pipeline_models", {})
                meta = models.get("metadata_extraction", {})
                vect = models.get("vector_embeddings", {})
                
                if meta.get("provider"):
                    config["llm_provider"] = meta["provider"]
                if meta.get("model"):
                    config["llm_model"] = meta["model"]
                if vect.get("provider"):
                    config["embedding_provider"] = vect["provider"]
                if vect.get("model"):
                    config["embedding_model"] = vect["model"]
        except Exception as e:
            print(f"Warning: Failed to load manifest: {e}")
            
    # Environmental overrides
    for k in config.keys():
        env_val = os.getenv(k.upper())
        if env_val:
            config[k] = env_val
            
    return config

def log_forensic_response(provider, model, prompt, system_prompt, response_text, reasoning_content=None, config=None):
    log_path = config.get("forensic_log_path", "FORENSIC_LOG") if config else "FORENSIC_LOG"
    timestamp = time.strftime("%Y-%m-%d %H:%M:%S")
    
    sys_prompt_str = f"SYSTEM PROMPT:\n{system_prompt}\n" if system_prompt else ""
    
    log_entry = (
        "================================================================================\n"
        f"TIMESTAMP: {timestamp}\n"
        f"PROVIDER:  {provider}\n"
        f"MODEL:     {model}\n"
        f"{sys_prompt_str}"
        "------------------------------------ PROMPT ------------------------------------\n"
        f"{prompt}\n"
        "---------------------------------- RESPONSE ------------------------------------\n"
    )
    
    if reasoning_content:
        log_entry += (
            "--- THINKING PROCESS ---\n"
            f"{reasoning_content}\n"
            "--- FINAL ANSWER ---\n"
        )
        
    log_entry += (
        f"{response_text}\n"
        "================================================================================\n\n"
    )
    
    try:
        with open(log_path, "a") as f:
            f.write(log_entry)
    except Exception as e:
        print(f"Warning: Failed to write to FORENSIC_LOG: {e}")

def generate_completion(prompt, system_prompt=None, json_mode=False, config=None):
    if not config:
        config = get_config()
        
    provider = config.get("llm_provider", "openrouter").lower()
    model = config.get("llm_model", "liquid/lfm-2-24b-a2b")
    
    messages = []
    if system_prompt:
        messages.append({"role": "system", "content": system_prompt})
    messages.append({"role": "user", "content": prompt})
    
    max_retries = 3
    attempt = 1
    
    while attempt <= max_retries:
        content = None
        reasoning = None
        err_msg = None
        
        try:
            with httpx.Client(timeout=60.0) as client:
                if provider == "openrouter":
                    api_key = config.get("openrouter_api_key")
                    if not api_key:
                        raise ValueError("openrouter_api_key is not configured.")
                        
                    headers = {
                        "Authorization": f"Bearer {api_key}",
                        "HTTP-Referer": "https://github.com/Arf9999/newsletter_phase2",
                        "X-Title": "Narrative Intelligence Pipeline"
                    }
                    body = {
                        "model": model,
                        "messages": messages,
                        "max_tokens": 1500
                    }
                    if json_mode:
                        body["response_format"] = {"type": "json_object"}
                        
                    resp = client.post("https://openrouter.ai/api/v1/chat/completions", json=body, headers=headers)
                    if resp.status_code != 200:
                        raise Exception(f"OpenRouter error: {resp.text}")
                        
                    res_json = resp.json()
                    choice_msg = res_json["choices"][0]["message"]
                    content = choice_msg.get("content")
                    reasoning = choice_msg.get("reasoning_content") or choice_msg.get("reasoning")
                    
                elif provider == "ollama":
                    url = f"{config.get('ollama_host', 'http://localhost:11434')}/api/chat"
                    body = {
                        "model": model,
                        "messages": messages,
                        "stream": False
                    }
                    if json_mode:
                        body["format"] = "json"
                        
                    resp = client.post(url, json=body)
                    if resp.status_code != 200:
                        raise Exception(f"Ollama error: {resp.text}")
                        
                    res_json = resp.json()
                    content = res_json["message"]["content"]
                    
                elif provider == "openai":
                    api_key = config.get("openai_api_key") or config.get("openrouter_api_key")
                    if not api_key:
                        raise ValueError("openai_api_key is not configured.")
                        
                    headers = {"Authorization": f"Bearer {api_key}"}
                    body = {
                        "model": model,
                        "messages": messages
                    }
                    if json_mode:
                        body["response_format"] = {"type": "json_object"}
                        
                    resp = client.post("https://api.openai.com/v1/chat/completions", json=body, headers=headers)
                    if resp.status_code != 200:
                        raise Exception(f"OpenAI error: {resp.text}")
                        
                    res_json = resp.json()
                    choice_msg = res_json["choices"][0]["message"]
                    content = choice_msg.get("content")
                    reasoning = choice_msg.get("reasoning_content")
                    
                else:
                    raise ValueError(f"Unknown LLM provider: {provider}")
                    
        except Exception as e:
            err_msg = str(e)
            
        if err_msg:
            print(f"LLM API call failed on attempt {attempt}: {err_msg}")
            attempt += 1
            if attempt <= max_retries:
                time.sleep(1)
            continue
            
        if content:
            # Check for recurring characters (e.g. repeated words/symbols)
            # Match R regex: grepl("(.{3,})\\1{8,}", content, perl = TRUE)
            has_repeats = bool(re.search(r'(.{3,})\1{8,}', content))
            json_invalid = False
            
            if json_mode:
                try:
                    json.loads(content)
                except Exception:
                    json_invalid = True
                    
            if has_repeats or json_invalid:
                reason = "recurring characters detected" if has_repeats else "malformed JSON/truncated"
                print(f"LLM output rejected ({reason}) on attempt {attempt}. Retrying...")
                
                log_forensic_response(
                    provider=f"{provider}-rejected-{reason}",
                    model=model,
                    prompt=prompt,
                    system_prompt=system_prompt,
                    response_text=content,
                    reasoning_content=reasoning,
                    config=config
                )
                attempt += 1
                if attempt <= max_retries:
                    time.sleep(1)
            else:
                log_forensic_response(
                    provider=provider,
                    model=model,
                    prompt=prompt,
                    system_prompt=system_prompt,
                    response_text=content,
                    reasoning_content=reasoning,
                    config=config
                )
                return content
        else:
            print(f"Empty response on attempt {attempt}. Retrying...")
            attempt += 1
            if attempt <= max_retries:
                time.sleep(1)
                
    raise Exception(f"Failed to obtain a valid completion after {max_retries} attempts.")

def generate_embeddings(text, config=None):
    if not config:
        config = get_config()
        
    provider = config.get("embedding_provider", "openrouter").lower()
    model = config.get("embedding_model", "nomic-embed-text")
    
    if not text or not text.strip():
        return None
        
    with httpx.Client(timeout=30.0) as client:
        if provider == "openrouter" or provider == "openai":
            api_key = config.get("openrouter_api_key") or config.get("openai_api_key")
            if not api_key:
                raise ValueError("No API key configured for embeddings.")
                
            headers = {"Authorization": f"Bearer {api_key}"}
            body = {
                "model": model,
                "input": text
            }
            resp = client.post("https://api.openai.com/v1/embeddings", json=body, headers=headers)
            if resp.status_code != 200:
                raise Exception(f"Embeddings error: {resp.text}")
            res_json = resp.json()
            return res_json["data"][0]["embedding"]
            
        elif provider == "ollama":
            url = f"{config.get('ollama_host', 'http://localhost:11434')}/api/embed"
            body = {
                "model": model,
                "input": text
            }
            resp = client.post(url, json=body)
            if resp.status_code != 200:
                raise Exception(f"Ollama error: {resp.text}")
            res_json = resp.json()
            return res_json["embeddings"][0]
            
        else:
            raise ValueError(f"Unknown embedding provider: {provider}")
